File: core/src/main/scala/com/graphbrain/eco/nodes/WWRecursion.scala
package com.graphbrain.eco.nodes

import com.graphbrain.eco.{Contexts, NodeType}
import com.graphbrain.eco.NodeType.NodeType

class WWRecursion(params: Array[ProgNode], lastTokenPos: Int= -1) extends FunNode(params, lastTokenPos) {
  override val label = ":ww"

  override def ntype: NodeType = NodeType.Words

  override def wordsValue(ctxts: Contexts) = {
    params(0).wordsValue(ctxts)
    for (c <- ctxts.ctxts) {
      val newCtxts = ctxts.prog.ww(c.getRetWords(params(0)))

      for (nctxts <- newCtxts) {
        for (nc <- nctxts.ctxts) {
          val forkCtxt = c.clone()
          forkCtxt.setRetWords(this, nc.getTopRetWords)
          forkCtxt.addSubContext(nc)

          // add forked context to caller contexts
          ctxts.addContext(forkCtxt)
        }
      }

      // remove original context
      ctxts.remContext(c)
    }
    ctxts.applyChanges()
  }
}

File: core/src/main/java/com/graphbrain/eco/nodes/WWRecursion.java
package com.graphbrain.eco.nodes;

import com.graphbrain.eco.Context;
import com.graphbrain.eco.Contexts;
import com.graphbrain.eco.NodeType;

import java.util.List;

public class WWRecursion extends FunNode {

    public WWRecursion(ProgNode[] params, int lastTokenPos) {
        super(params, lastTokenPos);
    }

    public WWRecursion(ProgNode[] params) {
        this(params, -1);
    }

    @Override
    public String label() {return ":ww";}

    @Override
    public NodeType ntype(Context ctxt) {return NodeType.Words;}

    @Override
    public void eval(Contexts ctxts) {
        params[0].eval(ctxts);
        for (Context c : ctxts.getCtxts()) {
            List<Contexts> newCtxts = ctxts.getProg().ww(c.getRetWords(params[0]), ctxts.getDepth() + 1, c);

            for (Contexts nctxts : newCtxts) {
                for (Context nc : nctxts.getCtxts()) {
                    Context forkCtxt = c.copy();
                    forkCtxt.setRetWords(this, nc.getTopRetWords());
                    forkCtxt.addSubContext(nc);

                    // add forked context to caller contexts
                    ctxts.addContext(forkCtxt);
                }
            }

            // remove original context
            ctxts.remContext(c);
        }
        ctxts.applyChanges();
    }
}
